File: backend/models/document.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type DocumentStatus string
type DocumentPriority int

const (
	StatusPending  DocumentStatus = "pending"
	StatusApproved DocumentStatus = "approved"
	StatusRejected DocumentStatus = "rejected"
	StatusExpired  DocumentStatus = "expired"
)

const (
	PriorityLow    DocumentPriority = 1
	PriorityMedium DocumentPriority = 2
	PriorityHigh   DocumentPriority = 3
)

type Document struct {
	ID              uint             `gorm:"primaryKey" json:"id"`
	Title           string           `gorm:"size:255;not null" json:"title"`
	Description     string           `gorm:"type:text" json:"description"`
	FilePath        string           `gorm:"size:500" json:"file_path"`
	Priority        DocumentPriority `gorm:"default:1" json:"priority"`
	Status          DocumentStatus   `gorm:"size:50;default:'pending'" json:"status"`
	RejectionReason string           `gorm:"type:text" json:"rejection_reason,omitempty"`
	Deadline        *time.Time       `json:"deadline,omitempty"`
	CreatorID       uint             `gorm:"not null" json:"creator_id"`
	AssignedToID    *uint            `json:"assigned_to_id,omitempty"`
	CreatedAt       time.Time        `json:"created_at"`
	UpdatedAt       time.Time        `json:"updated_at"`
	DeletedAt       gorm.DeletedAt   `gorm:"index" json:"-"`

	// Relations
	Creator    User      `gorm:"foreignKey:CreatorID" json:"creator,omitempty"`
	AssignedTo *User     `gorm:"foreignKey:AssignedToID" json:"assigned_to,omitempty"`
	History    []History `gorm:"foreignKey:DocumentID" json:"history,omitempty"`
}

type DocumentResponse struct {
	ID              uint             `json:"id"`
	Title           string           `json:"title"`
	Description     string           `json:"description"`
	FilePath        string           `json:"file_path"`
	Priority        DocumentPriority `json:"priority"`
	Status          DocumentStatus   `json:"status"`
	RejectionReason string           `json:"rejection_reason,omitempty"`
	Deadline        *time.Time       `json:"deadline,omitempty"`
	CreatorID       uint             `json:"creator_id"`
	AssignedToID    *uint            `json:"assigned_to_id,omitempty"`
	CreatorName     string           `json:"creator_name,omitempty"`
	AssignedToName  string           `json:"assigned_to_name,omitempty"`
	CreatedAt       time.Time        `json:"created_at"`
	UpdatedAt       time.Time        `json:"updated_at"`
}

func (d *Document) ToResponse() DocumentResponse {
	resp := DocumentResponse{
		ID:              d.ID,
		Title:           d.Title,
		Description:     d.Description,
		FilePath:        d.FilePath,
		Priority:        d.Priority,
		Status:          d.Status,
		RejectionReason: d.RejectionReason,
		Deadline:        d.Deadline,
		CreatorID:       d.CreatorID,
		AssignedToID:    d.AssignedToID,
		CreatedAt:       d.CreatedAt,
		UpdatedAt:       d.UpdatedAt,
	}

	if d.Creator.ID != 0 {
		resp.CreatorName = d.Creator.FullName
	}

	if d.AssignedTo != nil && d.AssignedTo.ID != 0 {
		resp.AssignedToName = d.AssignedTo.FullName
	}

	return resp
}

File: frontend/lib/models/document.dart
class Document {
  final int id;
  final String title;
  final String description;
  final String filePath;
  final int priority;
  final String status;
  final String? rejectionReason;
  final DateTime? deadline;
  final int creatorId;
  final int? assignedToId;
  final String? creatorName;
  final String? assignedToName;
  final DateTime createdAt;
  final DateTime updatedAt;

  Document({
    required this.id,
    required this.title,
    required this.description,
    required this.filePath,
    required this.priority,
    required this.status,
    this.rejectionReason,
    this.deadline,
    required this.creatorId,
    this.assignedToId,
    this.creatorName,
    this.assignedToName,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Document.fromJson(Map<String, dynamic> json) {
    return Document(
      id: json['id'] ?? 0,
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      filePath: json['file_path'] ?? '',
      priority: json['priority'] ?? 1,
      status: json['status'] ?? 'pending',
      rejectionReason: json['rejection_reason'],
      deadline: json['deadline'] != null 
          ? DateTime.parse(json['deadline']) 
          : null,
      creatorId: json['creator_id'] ?? 0,
      assignedToId: json['assigned_to_id'],
      creatorName: json['creator_name'],
      assignedToName: json['assigned_to_name'],
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at']) 
          : DateTime.now(),
      updatedAt: json['updated_at'] != null 
          ? DateTime.parse(json['updated_at']) 
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'file_path': filePath,
      'priority': priority,
      'status': status,
      'rejection_reason': rejectionReason,
      'deadline': deadline?.toIso8601String(),
      'creator_id': creatorId,
      'assigned_to_id': assignedToId,
      'creator_name': creatorName,
      'assigned_to_name': assignedToName,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  bool get hasFile => filePath.isNotEmpty;
  bool get isPending => status == 'pending';
  bool get isApproved => status == 'approved';
  bool get isRejected => status == 'rejected';
  bool get isExpired => status == 'expired';

  String get priorityLabel {
    switch (priority) {
      case 3:
        return 'High';
      case 2:
        return 'Medium';
      default:
        return 'Low';
    }
  }

  String get statusLabel {
    switch (status) {
      case 'approved':
        return 'Approved';
      case 'rejected':
        return 'Rejected';
      case 'expired':
        return 'Expired';
      default:
        return 'Pending';
    }
  }
}
